plugins/docsettingtweak.koplugin/main.lua

```diff
@@ -0,0 +1,106 @@
+end
+
+function DocSettingTweak:addToMainMenu(menu_items)
+    menu_items.doc_setting_tweak = {
+        text = _("Tweak document settings"),
+        callback = function() DocSettingTweak:editDirectoryDefaults() end,
+    }
+end
+
+function DocSettingTweak:lazyInitialization()
+    if not initialized then
+        if not lfs.attributes(directory_defaults_path, "mode") then
+            FFIUtil.copyFile(FFIUtil.joinPath(self.path, directory_defaults_name),
+                         directory_defaults_path)
+        end
+        initialized = true
+    end
+    DocSettingTweak:loadDefaults()
+end
+
+function DocSettingTweak:editDirectoryDefaults()
+    local directory_defaults_file = io.open(directory_defaults_path, "rb")
+    local defaults = directory_defaults_file:read("*all")
+    directory_defaults_file:close()
+    local config_editor
+    config_editor = InputDialog:new{
+        title = T(_("Directory Defaults: %1"),directory_defaults_path),
+        input = defaults,
+        input_type = "string",
+        fullscreen = true,
+        condensed = true,
+        allow_newline = true,
+        cursor_at_end = false,
+        add_nav_bar = true,
+        reset_callback = function()
+            return defaults
+        end,
+        save_callback = function(content)
+            if content and #content > 0 then
+                local parse_error = util.checkLuaSyntax(content)
+                if not parse_error then
+                    local syntax_okay, syntax_error = pcall(loadstring(content))
+                    if syntax_okay then
+                        directory_defaults_file = io.open(directory_defaults_path, "w")
+                        directory_defaults_file:write(content)
+                        directory_defaults_file:close()
+                        DocSettingTweak:loadDefaults()
+                        return true, _("Defaults saved")
+                    else
+                        return false, T(_("Defaults invalid: %1"), syntax_error)
+                    end
+                else
+                        return false, T(_("Defaults invalid: %1"), parse_error)
+                    end
+            end
+            return false, _("Defaults empty")
+        end,
+    }
+    UIManager:show(config_editor)
+    config_editor:onShowKeyboard()
+end
+
+function DocSettingTweak:onDocSettingsLoad(doc_settings, docfile)
+    if next(doc_settings.data) == nil and directory_defaults.data ~= nil then
+        local base = G_reader_settings:readSetting("home_dir") or filemanagerutil.getDefaultDir()
+        local directory = docfile and docfile ~= "" and FFIUtil.dirname(docfile)
+        while directory:sub(1, #base) == base do
+            if directory_defaults:has(directory) then
+                doc_settings.data = util.tableDeepCopy(directory_defaults:readSetting(directory))
+                break
+            else
+                directory = FFIUtil.dirname(directory)
+            end
+        end
+    end
+    return true
```

Comment:
Indeed, stopping propagation here largely defeats the purpose.

Reply:
i had thought that best practice was to always return true from an event, however i see how this can be used by others.

Reply:
We're conceptually talking about a generic hook here. Perhaps "hooks" and "events" could be separated, though I'm not currently inclined to think it'd make much of a practical difference in terms of real-world clarity or ease of use.

> i had thought that best practice was to always return true from an event

I would argue that requiring boilerplate code to reach a desired default state of affairs has appearances going against it.

But of course with a user input event, you'll almost certainly want to stop further propagation. Otherwise your on-screen button press would also turn the page, or something along those lines.